Omit document referrer should adopt the "network connection integrity" setting

charliewolfe opened the pull request:
#### 0ca9dcfa277a15c0d6ff789ab762c8703deb8847
<pre>
Omit document referrer should adopt the &quot;network connection integrity&quot; setting
<a href="https://bugs.webkit.org/show_bug.cgi?id=247627">https://bugs.webkit.org/show_bug.cgi?id=247627</a>
rdar://102108935

Reviewed by Wenson Hsieh.

Adopt the network connection integrity setting and remove the check for
ephemeral mode when omitting referrer for bindings. Also add a test option
for this setting and update layout tests.

* LayoutTests/http/tests/resourceLoadStatistics/omit-document-referrer-nested-third-party-iframe-ephemeral.html:
* LayoutTests/http/tests/resourceLoadStatistics/omit-document-referrer-third-party-iframe-ephemeral.html:
* LayoutTests/http/tests/resourceLoadStatistics/omit-referrer-for-navigation-ephemeral.html:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::referrerForBindings):
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::networkConnectionIntegrityEnabled const):
* Tools/WebKitTestRunner/TestController.h
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::configureContentMode):

Canonical link: <a href="https://commits.webkit.org/256548@main">https://commits.webkit.org/256548@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74643b2d7a91ba5b52118dd81b18524cfd93b471

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96042 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5291 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29084 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105593 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165916 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100020 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5404 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34062 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88420 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101410 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101702 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3994 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82646 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31025 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87747 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73860 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39783 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37459 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20618 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4529 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42055 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43225 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43945 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39881 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
```diff
@@ -555,6 +556,7 @@ static WKContentMode contentMode(const TestOptions& options)
 void TestController::configureContentMode(WKWebViewConfiguration *configuration, const TestOptions& options)
```
Nit - we probably want to rename `configureContentMode` now to something like `configureWebpagePreferences`, to reflect the fact that it's setting both bits.

whsieh commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5455,8 +5455,8 @@ String Document::referrer()
 {
-    if (auto* page = this->page(); page
-        && page->usesEphemeralSession()
+    if (auto* loader = this->topDocument().loader(); loader
```
Nit:
```suggestion
    if (auto* loader = topDocument().loader(); loader
```

whsieh reviewed (approved):
r=mews (with a couple of minor comments)

charliewolfe committed:
Omit document referrer should adopt the "network connection integrity" setting
https://bugs.webkit.org/show_bug.cgi?id=247627
rdar://102108935

Reviewed by Wenson Hsieh.

Adopt the network connection integrity setting and remove the check for
ephemeral mode when omitting referrer for bindings. Also add a test option
for this setting and update layout tests.

* LayoutTests/http/tests/resourceLoadStatistics/omit-document-referrer-nested-third-party-iframe-ephemeral.html:
* LayoutTests/http/tests/resourceLoadStatistics/omit-document-referrer-third-party-iframe-ephemeral.html:
* LayoutTests/http/tests/resourceLoadStatistics/omit-referrer-for-navigation-ephemeral.html:
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::referrerForBindings):
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::networkConnectionIntegrityEnabled const):
* Tools/WebKitTestRunner/TestController.h
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::configureContentMode):

Canonical link: https://commits.webkit.org/256548@main
